File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        
        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        
        ListNode t1 = head;
        ListNode t2 = slow.next;
        // Assigining the slow.next -> null
        slow.next = null;
        
        t2 = reverseLL(t2);
        
        ListNode dummy = new ListNode(-1);
        ListNode ls = dummy;
        while(t1 != null && t2 != null){
            ls.next = t1;
            ls = t1;
            t1 = t1.next;
            
            ls.next = t2;
            ls = t2;
            t2 = t2.next;
        }
        // If any LL left:
        if(t1 != null){
            ls.next = t1;
        }
        else if(t2 != null){
            ls.next = t2;
        }
        
    }
    public ListNode reverseLL(ListNode head){
        ListNode curr = head;
        ListNode prev = null;
        while(curr != null){
            ListNode nextNode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }
        return prev;
    }
}

File: 0143-reorder-list/0143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    let fast = head;
    let slow = head;
    
    // Finding the middle Node:
    while(fast && fast.next){
        fast = fast.next.next;
        slow = slow.next;
    }
    
    // Assigning two new LL:
    let t1 = head;
    let t2 = slow.next;
    
    // Making slow.next -> null.
    slow.next = null;
    
    t2 = reverseLL(t2);
    
    let d1 = new ListNode(-1);
    let l1 = d1;
    
    // Forming the LL:
    while(t1 && t2){
        l1.next = t1;
        l1 = t1;
        t1 = t1.next;
        
        l1.next = t2;
        l1 = t2;
        t2 = t2.next;
    }
    
    // If any LL left:
    if(t1){
        l1.next = t1;
    }
    if(t2){
        l1.next = t2;
    }
    
    return d1.next;
};

let reverseLL = (head) => {
    let curr = head;
    let prev = null;
    while(curr){
        let nextNode = curr.next;
        curr.next = prev;
        prev = curr;
        
        curr = nextNode;
    }
    
    return prev;
};